Implement our own changelog generator

vigneshsankariyer1234567890 opened the pull request:
# Description

Closes #2682.

Currently, we use Github's Changelog generator: https://docs.github.com/en/repositories/releasing-projects-on-github/automatically-generated-release-notes

However, this is not the best option since it increases reliance on Github actions themselves. Furthermore, there is a need for local changelog generation in order to simplify the release process.

In the initial issue, attempts were made to look for a tool that could generate the Changelogs directly. Some tools include [git-chglog](https://github.com/git-chglog/git-chglog), which actually generate the logs based on the commits themselves. However, these tools are not useful because FerretDB does not follow Changelog commit message standards at the moment, which leads to non-standard commit messages. This also means that it is difficult to use a tool that groups commit messages to hence get the list of PRs.

Instead, I opted for a custom approach that utilised the existing changelog template found in FerretDB (under .github/release.yml). Instead of using commits directly, the idea is to get the merged PRs from the Github API endpoint after the latest tag (which indicates a release). With the definition provided in the release.yml, we can emulate the Changelog generation action locally by subsequently grouping the PRs by the labels defined in the template, and lastly rendering as markdown to stdout. 

Here is a demo of how it works:

https://github.com/FerretDB/FerretDB/assets/68676914/9df77c0f-bcb2-41c5-980b-620799a3984d


To test it, generate binaries for tools as per instructions in the CONTRIBUTING.md, and launch task changelog directly from root folder, with an input MILESTONE_TITLE according to the milestone that you wish to generate the Changelog for. 

Closes #2682 .

## Readiness checklist

- [x] I added/updated unit tests (and they pass).
- [x] I added/updated integration/compatibility tests (and they pass).
- [x] I added/updated comments and checked rendering.
- [x] I made spot refactorings.
- [x] I updated user documentation.
- [x] I ran `task all`, and it passed.
- [x] I ensured that PR title is good enough for the changelog.
- [x] (for maintainers only) I set Reviewers ([`@FerretDB/core`](https://github.com/orgs/FerretDB/teams/core)), Milestone (`Next`), Labels, Project and project's Sprint fields.
- [x] I marked all done items in this checklist.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/FerretDB/FerretDB?pullRequest=4219) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/FerretDB/FerretDB?pullRequest=4219) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/FerretDB/FerretDB?pullRequest=4219) it.</sub>

AlekSi commented on tools/go.work:
Please remove that file

vigneshsankariyer1234567890 replied:
removed workspace file

AlekSi commented on tools/generatechangelog/go.mod:
Please move that into `tools/go.mod`

vigneshsankariyer1234567890 replied:
moved into `tools/go.mod`

AlekSi replied:
First of all: https://github.com/FerretDB/FerretDB/blob/c299c531fd40dfae5da03b241231ea480736f086/CONTRIBUTING.md?plain=1#L359-L363

Second, `tools/go.mod` wasn't updated.
Third, CI fails: https://github.com/FerretDB/FerretDB/actions/runs/8506940135/job/23298489403?pr=4219. Please run formatting, linting, etc locally as described in CONTRIBUTING.md

vigneshsankariyer1234567890 replied:
Apologies regarding comment resolution.

I have updated `tools/go.mod` to include the dependency for `yaml.v3` (and amended formatting issues)

AlekSi commented on tools/generatechangelog/generatechangelog.go:
```diff
@@ -0,0 +1,282 @@
+	"text/template"
+	"time"
+
+	"github.com/go-git/go-git/v5"
```
We should use GitHub's API instead for everything. See https://github.com/FerretDB/gh for some helpers, and please use `github.com/google/go-github/v57/github` for the rest

vigneshsankariyer1234567890 replied:
refactored to use `go-github`

AlekSi commented on tools/generatechangelog/generatechangelog_test.go:
```diff
@@ -0,0 +1,322 @@
+)
+
+const (
+	TAG_NAME = "v1.21.0"
```
https://go.dev/doc/effective_go#mixed-caps

AlekSi commented on tools/generatechangelog/generatechangelog_test.go:
```diff
@@ -0,0 +1,322 @@
+
+func TestGetLatestTag(t *testing.T) {
+	cwd, err := os.Getwd()
+	if err != nil {
+		t.Fatalf("Failed to get current working directory: %v", err)
+	}
```
```suggestion

	require.NoError(t, err)
```

We use testify for all tests

vigneshsankariyer1234567890 replied:
refactored to use `testify`

AlekSi commented:
@vigneshsankariyer1234567890 I will do a re-review shortly; please see our contribution documentation in the meantime

AlekSi commented on tools/go.mod:
```diff
@@ -20,6 +20,8 @@ require (
 	mvdan.cc/gofumpt v0.6.0
 )
 
+require gopkg.in/yaml.v3 v3.0.1
```
Please move it to the group above

vigneshsankariyer1234567890 replied:
Added to group as requested, and run `task lint` to verify formatting.

AlekSi commented on tools/tools.go:
```diff
@@ -25,11 +25,14 @@ import (
 	_ "golang.org/x/vuln/cmd/govulncheck"
 	_ "mvdan.cc/gofumpt"
+
+	_ "github.com/FerretDB/FerretDB/tools/generatechangelog"
```
It should not be there – note that checkdocs, checkswitch, checkcomments are all not imported

vigneshsankariyer1234567890 replied:
Removed verified binary generation still works.

vigneshsankariyer1234567890 commented:
@AlekSi regarding tests for the introduced tool, I noticed that `tools/Taskfile.yml` has a script for testing `checkcomments`, `checkdocs`, `checkswitch` etc. Should I also be adding the test for `generatechangelog` within that?

AlekSi commented:
Yes, absolutely. Please follow existing things and style as close as possible

vigneshsankariyer1234567890 commented:
@AlekSi requested your review, and ensured lint and formatting passes. At the moment, running `task all` fails due to internal test failures, but I'm checking if it is an issue with my device.

vigneshsankariyer1234567890 commented on tools/generatechangelog/generatechangelog.go:
```diff
@@ -0,0 +1,245 @@
+	query := fmt.Sprintf("repo:%s/%s is:pr is:merged merged:>=%s", owner, repo, sinceDate)
+	opts := &github.SearchOptions{Sort: "created", Order: "desc", ListOptions: github.ListOptions{PerPage: 100}}
+
+	issues, _, err := client.Search.Issues(ctx, query, opts)
```
Understood. Just to be clear, this would be the flow:

1. Get the milestone based on title
2. Search for PRs which are closed and merged based on the milestone number
3. Format according to release template

AlekSi replied:
Re 2 – not “search”. Search is a different and more complicated API. We just need to get a list of PRs within the milestone: https://github.com/FerretDB/FerretDB/milestone/63?closed=1

vigneshsankariyer1234567890 replied:
Understood. I used the "Issues" API instead, which has a `ListByRepo` method which allows to pass the milestone number and filter by closed. Subsequently, I identified Issues as Pull Requests by using the `Issue::IsPullRequest` method. However, I'm not sure about the distinction between Closed and Merged Pull requests, so I will take a further look at the API to clarify this.

vigneshsankariyer1234567890 commented on tools/generatechangelog/changelog_template.tmpl:
```diff
@@ -0,0 +1,8 @@
+- {{ .Title }} by @{{ .User.Login }} in {{ .URL }}
+{{- end }}
+{{- "\n" }}
+{{- end }}
```
added line break

vigneshsankariyer1234567890 commented on tools/generatechangelog/generatechangelog.go:
```diff
@@ -0,0 +1,245 @@
+	"gopkg.in/yaml.v3"
+)
+
+const (
+	OrgOwner = "FerretDB"
+	Repo     = "FerretDB"
+)
```
removed constants, and used directly in github call

vigneshsankariyer1234567890 commented on tools/generatechangelog/generatechangelog.go:
```diff
@@ -0,0 +1,245 @@
+	return &commit.Author.Date.Time, nil
+}
+
+func FetchPRs(ctx context.Context, client *github.Client, owner, repo, sinceDate string) ([]PRItem, error) {
```
removed params for owner and repo

vigneshsankariyer1234567890 commented on tools/generatechangelog/generatechangelog.go:
```diff
@@ -0,0 +1,245 @@
+	res := []PRItem{}
+
+	for _, issue := range issues.Issues {
+		res = append(res, func(i *github.Issue) PRItem {
```
Removed the JSON tags, they were there originally for verification

vigneshsankariyer1234567890 commented on tools/generatechangelog/generatechangelog.go:
```diff
@@ -0,0 +1,245 @@
+	}
+
+	var template *ReleaseTemplate
+	err = yaml.Unmarshal(bytes, &template)
```
amended template

vigneshsankariyer1234567890 commented on tools/generatechangelog/generatechangelog_test.go:
```diff
@@ -0,0 +1,216 @@
+func TestGroupPRsByCategories(t *testing.T) {
+	prItems := []PRItem{
+		{
+			URL:    "http://example.com/pr1",
```
I've added an integration test that does each individual part and checks based on the output, which has real PRs.

vigneshsankariyer1234567890 commented:
@AlekSi requested for your re-review. As requested, I have changed it such that `task changelog` takes in a variable "MILESTONE_TITLE", and can be used like this: `task changelog MILESTONE_TITLE="your_milestone_here"`. I have also amended it to use the Issues API directly, and added an integration test that checks if the output matches the expected PRs for an old version (v0.9.1).

rumyantseva commented on tools/generatechangelog/generatechangelog_test.go:
```diff
@@ -0,0 +1,196 @@
+
+func TestGetMilestone(t *testing.T) {
+	ctx := context.Background()
+	client, err := gh.NewRESTClient(os.Getenv("GITHUB_TOKEN"), nil)
```
As the tests (this one and the one below) work with real GitHub API, ideally, we need to set `GITHUB_TOKEN` on CI when running tests. Otherwise, we might be rate-limited.

What do you think?

AlekSi commented:
Thank you!
